lib/src/main/java/org/esbtools/lightbluenotificationhook/NotificationHook.java

```diff
@@ -110,60 +134,95 @@ private HookResult processSingleHookDoc(Projector watchProjector, Projector incl
         return HookResult.aborted();
     }
 
-    private boolean watchedFieldsHaveChanged(JsonDoc preDoc, JsonDoc postDoc,
-            Projector watchProjector) {
+    /**
+     * Determines if any of the watched fields have changed. If
+     * arrayOrderingSignificant is true, also checks if array
+     * orderings also changed.
+     */
+    private boolean watchedFieldsHaveChanged(EntityMetadata metadata,
+                                             JsonDoc preDoc,
+                                             JsonDoc postDoc,
+                                             Projector watchProjector,
+                                             boolean arrayOrderingSignificant) {
         JsonDoc watchedPostDoc = watchProjector.project(postDoc, jsonNodeFactory);
         JsonDoc watchedPreDoc = preDoc == null
-                ? new JsonDoc(jsonNodeFactory.nullNode())
-                : watchProjector.project(preDoc, jsonNodeFactory);
+            ? new JsonDoc(jsonNodeFactory.objectNode())
+            : watchProjector.project(preDoc, jsonNodeFactory);
 
-        return !watchedPostDoc.getRoot().equals(watchedPreDoc.getRoot());
+        // Compute diff
+        JsonCompare cmp=metadata.getDocComparator();
+        try {
+            DocComparator.Difference diff=cmp.compareNodes(watchedPreDoc.getRoot(),watchedPostDoc.getRoot());
+            LOGGER.debug("Diff: {}",diff);
+            if(!diff.same()) {
+                if(diff.getNumChangedFields()>0) {
+                    return true;
+                } else {
+                    // Are there any array element moves?
+                    // Do we care?
+                    if(!diff.getDelta().isEmpty()&&arrayOrderingSignificant) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        } catch (Exception e) {
+            LOGGER.error("Error processing notification",e);
+            return false;
+        }
     }
 
     private NotificationEntity makeNotificationEntityWithIncludedFields(HookDoc hookDoc,
-            Projector includeProjector) {
+                                                                        Projector includeProjector,
+                                                                        Projector propertiesProjector) {
         EntityMetadata metadata = hookDoc.getEntityMetadata();
         JsonDoc postDoc = hookDoc.getPostDoc();
 
-        List<NotificationEntity.PathAndValue> data = new ArrayList<>();
-
+        NotificationEntity notificationEntity = new NotificationEntity();
+        // Set the payload
+        JsonDoc includeDoc=includeProjector.project(postDoc,jsonNodeFactory);
+        notificationEntity.setEntityData(includeDoc.toString());
+        
+        List<NotificationEntity.FieldAndValue> properties=new ArrayList<>();
+        notificationEntity.setProperties(properties);
+        
+        // Set properties
+        if(propertiesProjector!=null) {
```

Comment:
Reverted most of the stuff back, this should address your concerns.

On Tue, Jan 26, 2016 at 12:05 PM, Alec Henninger <notifications@github.com>
wrote:

> In
> lib/src/main/java/org/esbtools/lightbluenotificationhook/NotificationHook.java
> <https://github.com/esbtools/lightblue-notification-hook/pull/1#discussion_r50882745>
> :
>
> >          EntityMetadata metadata = hookDoc.getEntityMetadata();
> >          JsonDoc postDoc = hookDoc.getPostDoc();
> >
> > -        List<NotificationEntity.PathAndValue> data = new ArrayList<>();
> > -
> > +        NotificationEntity notificationEntity = new NotificationEntity();
> > +        // Set the payload
> > +        JsonDoc includeDoc=includeProjector.project(postDoc,jsonNodeFactory);
> > +        notificationEntity.setEntityData(includeDoc.toString());
> > +
> > +        List<NotificationEntity.FieldAndValue> properties=new ArrayList<>();
> > +        notificationEntity.setProperties(properties);
> > +
> > +        // Set properties
> > +        if(propertiesProjector!=null) {
>
> In other words, in most use cases includedFields will be nothing, or
> include one or two fields. So I think we can get away with just using
> paths/fields and values to eliminate the redundancy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/esbtools/lightblue-notification-hook/pull/1/files#r50882745>
> .
>